### PR TITLE
Cartogears poc views

### PR DESCRIPTION
**Pull request by juanignaciosl**
This closes CartoDB/cartodb-central/issues/1701

### CODE REVIEW

File: config/initializers/carto_db.rb [hunk 1]
```diff
@@ -133,6 +138,11 @@ def self.protocol(protocol_override=nil)
   # "private" methods, not intended for direct usage
   # ------------------------------------------------
 
+  def self.organization_username(user)
+    subdomainless_urls? || user.organization.nil? ? nil : user.username
+  end
```
Comment (javitonino) [inline config/initializers/carto_db.rb]: Can you explain the reason to extract `organization_username` from the user model to here?

Best I can guess, it is somehow related to URL generation from the Gear, so I guess that the `CartoGearsApi::User` is being passed to the URL generation code, is that it?

File: gears/carto_gears_api/README.md [hunk 2]
```diff
@@ -1,6 +1,8 @@
 CARTO Gears API
 ===============
 
+**DISCLAIMER: CARTO Gears API is still in development, it won't be considered stable until first major 1 version**
```
Comment (javitonino) [inline gears/carto_gears_api/README.md]: The classic "You have been warned"

File: gears/carto_gears_api/README.md [hunk 3]
```diff
@@ -78,4 +80,22 @@ In order to enable rspec:
 
 # Documentation
 
+## `CartoGearsAPI::` (Ruby API)
 Generate the documentation with `yard`.
+
+## Extension points
+
+### Adding links to profile page
```
Comment (javitonino) [inline gears/carto_gears_api/README.md]: Nice example, I like that it includes how to get access to the user object. ❤️ 

I would move this to yard documentation. Not sure if under `gears/carto_gears_api/lib/carto_gears_api/pages/subheader.rb` or maybe we can create a "Getting Started" inside yard (not associated to any class). This has the benefit of linking classes to its yard doc page, and having everything in the same place. What do you think?

File: app/views/admin/shared/_pages_subheader.html.erb [hunk 4]
```diff
@@ -32,6 +32,9 @@
       <% if current_user.organization_owner? %>
         <li class="SideMenu-typeItem"><a href="<%= CartoDB.url(self, 'organization_settings', {}, current_user) %>" class="SideMenu-typeLink <%= 'is-selected' if ["admin/organizations", "admin/organization", "admin/organization_users"].include?("#{params[:controller]}") %>" class="SideMenu-typeLink">Organization settings</a></li>
       <% end %>
+      <% CartoGearsApi::Pages::Subheader.instance.links(self).each do |path, text, controller| %>
```
Comment (javitonino) [inline app/views/admin/shared/_pages_subheader.html.erb]: I'm thinking that the link generators should return an object of some kind instead of a fixed size array.

Pros:
- Easier to document the return code for the Gear programmer. We can even user keyword arguments to make it more explicit: `SubheaderLink.new(path: '', controller: '', text: '')`.
- Easier to extend. If we need to add another attribute for link generation, we can add it to the `SubheaderLink` object with a reasonable default, and we do not force Gear programmers to update the API.

Cons:
- More complexity that is not needed right now.
- Explosing of small classes (can be somewhat handled by declaring these classes inside the link generation singleton)

File: gears/carto_gears_api/lib/carto_gears_api/config.rb [hunk 5]
```diff
@@ -0,0 +1,7 @@
+module CartoGearsApi
+  class Config
+    def get_config(*config_chain)
```
Comment (javitonino) [inline gears/carto_gears_api/lib/carto_gears_api/config.rb]: I would love to take this chance to make this method take only symbols, and convert to symbol or string as needed. Our current `get_config` implementation takes the first one as a symbol and the rest as strings which is confusing to say the least. Let's make this easier for the Gear programmer.

To keep this focused, we can just to an adapter here (that takes symbols, and stringifies all of them except the first one), or a full reimplementation of the `get_config` method. We can also change the original `get_config` methods so it takes both parameters formats.

File: gears/carto_gears_api/lib/carto_gears_api/gear_controller_helper.rb [hunk 6]
```diff
@@ -0,0 +1,21 @@
+# This should be included in every controller using the following CARTO layouts:
```
Comment (javitonino) [inline gears/carto_gears_api/lib/carto_gears_api/gear_controller_helper.rb]: Nice comment. I think it's `should` -> `must`, as I think this is required and the layouts won't work without this.

File: gears/carto_gears_api/lib/carto_gears_api/http_server.rb [hunk 7]
```diff
@@ -0,0 +1,82 @@
+module CartoGearsApi
```
Comment (javitonino) [inline gears/carto_gears_api/lib/carto_gears_api/http_server.rb]: Trying to think from the perspective of a Gear programmer, I think this adds little value on top of a standard HTTP client. We should document why it's important to use this API instead of directly calling typhoeus or something like that:
- Logging
- No needed to add further dependencies
- ?

Why `HttpServer`? Why not `HttpClient` like the class it is wrapping?

File: gears/carto_gears_api/lib/carto_gears_api/url_helper.rb [hunk 8]
```diff
@@ -0,0 +1,17 @@
+module CartoGearsApi
+  module UrlHelper
+    # @param gear [Symbol] Gear name.
+    # @param context [ActionController::Base]
+    # @param path [String] path name defined at the current Gear.
+    def carto_gear_path(gear, context, path)
+      callable_context = context.respond_to?(gear) ? context.send(gear) : context
+      CartoDB.path(callable_context, path)
```
Comment (javitonino) [inline gears/carto_gears_api/lib/carto_gears_api/url_helper.rb]: Does `callable_context.main_app` resolve to the Gear? I would think it would resolve to the CartoDB application 😖

File: gears/carto_gears_api/lib/carto_gears_api/users_service.rb [hunk 9]
```diff
@@ -1,10 +1,42 @@
...
 module CartoGearsApi
   # User information.
   #
-  # @attr_reader [String]        email     Email
-  class User < Value.new(:email); end
+  # @attr_reader [UUID] id User id
+  # @attr_reader [String] username User name
+  # @attr_reader [String] email Email
+  # @attr_reader [CartoGearsApi::Organization] organization Organization
+  class User < Value.new(:id, :username, :email, :organization, :feature_flags)
```
Comment (javitonino) [inline gears/carto_gears_api/lib/carto_gears_api/users_service.rb]: Not important right now, since we can always make the change later.

How about passing the User model to this class, and implementing the attributes we want to expose through delegation? Alternatively, we could a factory that takes the model and builds the value object.

Just thinking out loud, these are internal API details and we can change them as needed, so not asking for changes.

File: gears/carto_gears_api/app/views/layouts/carto_gears_api/dashboard.html.erb [hunk 10]
```diff
@@ -0,0 +1,17 @@
+<%= content_for(:js) do -%>
```
Comment (javitonino) [inline gears/carto_gears_api/app/views/layouts/carto_gears_api/dashboard.html.erb]: I'm surprised this layout did not exist prior to this PR. 😕 

✨

File: app/controllers/carto/admin/user_public_map_adapter.rb [hunk 11]
```diff
@@ -5,7 +5,7 @@ class UserPublicMapAdapter
       extend Forwardable
 
       delegate [ :id, :name, :username, :disqus_shortname, :avatar, :avatar_url, :remove_logo?, :has_organization?, 
-                :organization, :organization_id, :twitter_username, :location, :public_url, :subdomain, :organization_username, 
+                :organization, :organization_id, :twitter_username, :location, :public_url, :subdomain,
```
Comment (juanignaciosl) [inline app/controllers/carto/admin/user_public_map_adapter.rb]: Not if it's removed from the models. More on this below.

File: app/helpers/application_helper.rb [hunk 12]
```diff
@@ -3,6 +3,14 @@
...
 module ApplicationHelper
   include CartoDB::ConfigUtils
+  include SafeJsObject
+  include TrackjsHelper
+  include GoogleAnalyticsHelper
+  include HubspotHelper
+  include FrontendConfigHelper
+  include AppAssetsHelper
+  include MapsApiHelper
+  include SqlApiHelper
```
Comment (juanignaciosl) [inline app/helpers/application_helper.rb]: No, nothing was needed. I extracted one by one as they were needed for the layouts, just cut and paste.

File: app/views/admin/shared/_pages_subheader.html.erb (hunk 4 above)
Comment (juanignaciosl) [inline app/views/admin/shared/_pages_subheader.html.erb]: Ok, I'll do a `CartoGearsApi::Pages::Link` class. I think that we can probably improve the needed parameters later (`controller` might not be needed), this abstraction will help with this 👍

File: config/initializers/carto_db.rb (hunk 1 above)
Comment (juanignaciosl) [inline config/initializers/carto_db.rb]: Yes, that's it. It's needed for URL generation, so I should've added it to `CartoGearsApi::User`. But, after a second look, it became obvious that it's only used here, so it doesn't belong to the model but to the generation itself.

File: gears/carto_gears_api/README.md (hunk 2 above)
Comment (juanignaciosl) [inline gears/carto_gears_api/README.md]: :D It's implicit at semver, but we're not explicitly declaring semver for this. We probably should (although I should study the details of it). What do you think?

File: gears/carto_gears_api/README.md (hunk 3 above)
Comment (juanignaciosl) [inline gears/carto_gears_api/README.md]: `README.md` is rendered at the YARD index page, which is pretty cool. Nevertheless, after the `SubheaderLink` refactor it's duplicated. I'll take a look at the proper way to avoid it 👍

File: gears/carto_gears_api/app/views/layouts/carto_gears_api/dashboard.html.erb (hunk 10 above)
Comment (juanignaciosl) [inline gears/carto_gears_api/app/views/layouts/carto_gears_api/dashboard.html.erb]: /shrug

File: gears/carto_gears_api/README.md (hunk 2 above)
Comment (javitonino) [inline gears/carto_gears_api/README.md]: Not a fan of semver, but it fits the Gears use case pretty well. IMHO, I would not even bother versioning CartoGearsAPI until we release the first Gear and have a more complete solution.

File: gears/carto_gears_api/README.md (hunk 3 above)
Comment (javitonino) [inline gears/carto_gears_api/README.md]: Oh, nice :) I should really install yard and compile the docs before talking 🙊

File: gears/carto_gears_api/README.md (hunk 2 above)
Comment (juanignaciosl) [inline gears/carto_gears_api/README.md]: No, versioning is not a problem yet (that's the reason for the disclaimer), I just wanted to prepare the discussion :-)

File: gears/carto_gears_api/lib/carto_gears_api/config.rb (hunk 5 above)
Comment (juanignaciosl) [inline gears/carto_gears_api/lib/carto_gears_api/config.rb]: Adapter approach 👍

File: gears/carto_gears_api/lib/carto_gears_api/url_helper.rb (hunk 8 above)
Comment (javitonino) [inline gears/carto_gears_api/lib/carto_gears_api/url_helper.rb]: ^ Unanswered, and I'm still curious.

Comment (juanignaciosl) [inline gears/carto_gears_api/lib/carto_gears_api/url_helper.rb]: Oh, sorry, I overlooked this one.

TL;DR: `callable_context` doesn't even respond to `main_app`. 

This is meant to be used in gear requests. `context.respond_to?(gear)` is `true` so `callable_ context` gets the value of `context.send(gear)`.
`callable_context` is a `ActionDispatch::Routing::RoutesProxy`. `main_app` is a helper that can be used in views and controllers, but not in the routes proxy needed here to get the path.

Comment (javitonino) [inline gears/carto_gears_api/lib/carto_gears_api/url_helper.rb]: Thanks for the explanation! 💖

File: gears/carto_gears_api/lib/carto_gears_api/http_server.rb (hunk 7 above)
Comment (juanignaciosl) [inline gears/carto_gears_api/lib/carto_gears_api/http_server.rb]: I overlooked this one as well, probably some commits hid both, please keep insisting if there are more unanswered questions.

I also had doubts about providing this or not, but your two points made me do it. I can remove it if you consider it shouldn't be there, though.

The naming is wrong. I chose `MyOtherApiServer` as the name in the gear and a is-a (inheritance) relationship felt natural, but even if I keep `MyOtherApiServer` naming, it shouldn't be inheritance but association. Changing...

File: gears/carto_gears_api/lib/carto_gears_api/users_service.rb (hunk 9 above)
Comment (javitonino) [inline gears/carto_gears_api/lib/carto_gears_api/users_service.rb]: ^ Oh, also found this comment as unanswered. It's just a thought, no action required. But in case you missed it.

File: gears/carto_gears_api/lib/carto_gears_api/http_server.rb (hunk 7 above)
Comment (javitonino) [inline gears/carto_gears_api/lib/carto_gears_api/http_server.rb]: > I overlooked this one as well, probably some commits hid both, please keep insisting if there are more unanswered questions.

Don't doubt it. You are not getting off that easy 😈 

> I also had doubts about providing this or not, but your two points made me do it. I can remove it if you consider it shouldn't be there, though.

It looked strange at first, but after thinking of the two points above, I think it is good so we can force everyone to use the same API if possible, and avoid adding extra dependencies to this gems for something so trivial as making HTTP requests. As long as we document this is the preferred approach for HTTP requests, I am alright with it.

File: gears/carto_gears_api/lib/carto_gears_api/http_server.rb [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+                                                 url: request.url, body: body, method: method, timeout: timeout,
+                                                 request: request, response: response)
+      end
+    rescue => e
+      CartoGearsApi::Logger.new('api').error(exception: e, additional_data: e.inspect)
+      raise e
+    end
+
+    def build_request(path:, body:, method:, content_type: { "Content-Type" => "application/json" }, timeout: 200)
+      http_client = Carto::Http::Client.get(@host, log_requests: true)
```
Comment (javitonino) [inline gears/carto_gears_api/lib/carto_gears_api/http_server.rb]: Idea: Maybe we can pass the gear name (instead/in addition to `@host`) to the client, so we know which gear is making the request.

File: gears/carto_gears_api/lib/carto_gears_api/users_service.rb (hunk 9 above)
Comment (juanignaciosl) [inline gears/carto_gears_api/lib/carto_gears_api/users_service.rb]: Because having the AR instance around is potentially dangerous (side effects because of callbacks), can leak connections... In addition, if we keep API models immutable everything will be easier. As soon as you wrap a mutable object, that's not longer true. I don't see a clear benefit.

**Comment by juanignaciosl** [comment]
Github is trolling me, your comments won't appear. I'll keep an eye on every mail notification.

> Idea: Maybe we can pass the gear name (instead/in addition to @host) to the client, so we know which gear is making the request.

As the trace contains the code paths and requests, I think that it isn't needed.

Waiting for 🍏  for a last ping and merge.

**Comment by javitonino** [comment]
> As the trace contains the code paths and requests, I think that it isn't needed.

Makes sense 👍

**Comment by javitonino** [comment]
retest this please

**Comment by juanignaciosl** [comment]
🍏  and deployment works 🎉 

@javitonino a last look before merging?

File: gears/carto_gears_api/app/views/shared/form/_input_text.html.erb [hunk 14]
```diff
@@ -0,0 +1,3 @@
+<%-# @param form [ActionView::Helpers::FormBuilder] Current form -%>
+<%-# @param field [Symbol] Model attribute for the field -%>
```
Comment (javitonino) [inline gears/carto_gears_api/app/views/shared/form/_input_text.html.erb]: I just discovered (writing another PR), that this is probably better as `<%# @param field [Symbol] Model attribute for the field %>` (erb comment), rather than `<%-#` (ruby evalution block with a ruby comment).

Don't know if it will work with yard then, and is very minor anyway. 
Also applies to more templates.

File: gears/carto_gears_api/lib/carto_gears_api/logger.rb [hunk 15]
```diff
@@ -4,7 +4,7 @@ module CartoGearsApi
   # Gears on CARTO should use this class for logging. It takes local directories and external
   # services conventions into account.
   class Logger
-    # @param gear_name The name of the gear. Log messages will be prefixed with `{gear_name}`.
+    # @param gear_name [String] The name of the gear. Log messages will be prefixed with it.
```
Comment (javitonino) [inline gears/carto_gears_api/lib/carto_gears_api/logger.rb]: Thinking about this, it would be cool to pass this as a custom variable to Rollbar as well, so it's even easier to filter:

Just changing a single line at the bottom should be enough. I would still keep the message prefix, makes for easier human reading, while the attribute is mostly for machine filtering.
`CartoDB::Logger.log(level, exception: exception, message: gear_message, user: user, gear: @gear, **additional_data)`

File: gears/carto_gears_api/app/views/shared/form/_input_text.html.erb (hunk 14 above)
Comment (juanignaciosl) [inline gears/carto_gears_api/app/views/shared/form/_input_text.html.erb]: First of all, could you provide more detail about how come `<%#` is _better_?

YARD support for ERBs is pretty much nonexisting. Nevertheless, after a couple of tests, this signature makes params _legible_:

```ruby
<%
  # @param form [ActionView::Helpers::FormBuilder] Current form
  # @param field [Symbol] Model attribute for the field
%>
```

captura de pantalla 2017-03-14 a las 12 27 51

What do you think?

File: gears/carto_gears_api/lib/carto_gears_api/logger.rb (hunk 15 above)
Comment (juanignaciosl) [inline gears/carto_gears_api/lib/carto_gears_api/logger.rb]: Cool ✨  On it.

File: gears/carto_gears_api/app/views/shared/form/_input_text.html.erb (hunk 14 above)
Comment (javitonino) [inline gears/carto_gears_api/app/views/shared/form/_input_text.html.erb]: I have no proof that `<%#` is better, but it should be. This is an erb comment, so it is ignored by the erb parser. Code inside `<%-` is passed through ruby's `eval` so it has to be parsed by a second interpreter, than then proceeds to ignore it. Anyway, I guess it doesn't matter much, so whatever works.

About yard, I guessed that it has to support based on these comments. I really really should run yard at least once (and this is the second time I'm saying this 😞). Do as you prefer, I don't see a huge advantage of having the file in yard (without linking) vs just looking at the file comments.

Comment (juanignaciosl) [inline gears/carto_gears_api/app/views/shared/form/_input_text.html.erb]: I think that I've just come out with a solution that has all that we need for now, take a look at 299a49e0075c87267187476c3503d06f10f7243c

Comment (javitonino) [inline gears/carto_gears_api/app/views/shared/form/_input_text.html.erb]: Yeah, it seems to work ok 👍 

Having to add them manually to the `yardoc` command is not great, but we can probable add some bash hack using `find` in the future if we feel it's useful.

**Comment by juanignaciosl** [comment]
Right! 🎉
